### PR TITLE
[mlir][LLVM] Add nneg flag

### PR DESCRIPTION
**Pull request by lfrenot**
This implementation is based on the existing one for the exact flag.

If the nneg flag is set and the argument is negative, the result is a poison value.

@zero9178, @gysit and @Dinistro, could you take a look?

### CODE REVIEW

File: mlir/include/mlir/Dialect/LLVMIR/LLVMOps.td [hunk 1]
```diff
@@ -508,6 +508,23 @@ class LLVM_CastOp<string mnemonic, string instName, Type type,
       $_location, $_resultType, $arg);
   }];
 }
+class LLVM_CastOpWithNNegFlag<string mnemonic, string instName, Type type,
+                  Type resultType, list<Trait> traits = []> :
+    LLVM_Op<mnemonic, !listconcat([Pure], [DeclareOpInterfaceMethods<NonNegFlagInterface>], traits)>,
+    LLVM_Builder<"$res = builder.Create" # instName # "($arg, $_resultType, /*Name=*/\"\", op.getNonNeg());"> {
+  let arguments = (ins type:$arg,  UnitAttr:$nonNeg);
```
Comment (gysit) [inline mlir/include/mlir/Dialect/LLVMIR/LLVMOps.td:515]:
```suggestion
  let arguments = (ins type:$arg, UnitAttr:$nonNeg);
```
ultra nit:

File: mlir/test/Dialect/LLVMIR/roundtrip.mlir [hunk 2]
```diff
@@ -325,6 +325,19 @@ func.func @casts(%arg0: i32, %arg1: i64, %arg2: vector<4xi32>,
...
 
+// CHECK-LABEL: @nneg_casts
+// CHECK-SAME: (%[[I32:.*]]: i32, %[[I64:.*]]: i64, %[[V4I32:.*]]: vector<4xi32>, %[[V4I64:.*]]: vector<4xi64>, %[[PTR:.*]]: !llvm.ptr)
+func.func @nneg_casts(%arg0: i32, %arg1: i64, %arg2: vector<4xi32>,
+                %arg3: vector<4xi64>, %arg4: !llvm.ptr) {
+// CHECK:  = llvm.zext nneg %[[I32]] : i32 to i64
+  %0 = llvm.zext nneg %arg0 : i32 to i64
+// CHECK:  = llvm.zext nneg %[[V4I32]] : vector<4xi32> to vector<4xi64>
+  %4 = llvm.zext nneg %arg2 : vector<4xi32> to vector<4xi64>
+// CHECK: %[[FLOAT:.*]] = llvm.uitofp nneg %[[I32]] : i32 to f32
```
Comment (gysit) [inline mlir/test/Dialect/LLVMIR/roundtrip.mlir:336]:
```suggestion
// CHECK:  = llvm.uitofp nneg %[[I32]] : i32 to f32
```
nit: I would drop the variable since it has no uses in this test.

File: mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp [hunk 3]
```diff
@@ -146,6 +146,9 @@ static void printLLVMOpAttrs(OpAsmPrinter &printer, Operation *op,
   } else if (auto iface = dyn_cast<ExactFlagInterface>(op)) {
     printer.printOptionalAttrDict(filteredAttrs,
                                   /*elidedAttrs=*/{iface.getIsExactName()});
+  } else if (auto iface = dyn_cast<NonNegFlagInterface>(op)) {
+    printer.printOptionalAttrDict(filteredAttrs,
```
Comment (gysit) [inline mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:150]: I suspect this is not needed since the operation does not use `custom<LLVMOpAttrs>(attr-dict)`. It seems like this print/parseLLVMOpAttrs should only be used for the operations that use fast math flags. That can be cleaned up in a separate PR though. 

I brief I would suggest to remove the NonNegFlagInterface here, except it is necessary for a reason I do not see right now?

**Review by gysit — APPROVED** [review]
Thanks!

LGTM modulo nit comments!

File: mlir/include/mlir/Dialect/LLVMIR/LLVMOps.td (hunk 1 above)
Comment (lfrenot) [reply mlir/include/mlir/Dialect/LLVMIR/LLVMOps.td]: Done!

File: mlir/test/Dialect/LLVMIR/roundtrip.mlir (hunk 2 above)
Comment (lfrenot) [reply mlir/test/Dialect/LLVMIR/roundtrip.mlir]: Done!

File: mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp (hunk 3 above)
Comment (lfrenot) [reply mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp]: Done!
